### PR TITLE
RTEMIS-188: Student Admission Report Page for States.

### PR DESCRIPTION
**Pull request by RohanSingh-inno**
Ticket # : https://innoraft.atlassian.net/browse/RTEMIS-188

---
### What the PR does
`(Explain fixes, changes and impacts)`
- Student Admission Report Page for States.
- Added a condition check for Allocation edit access check.
- Optimized the roleBasedDetailBlock.

---
### Extras
`(Add Screenshots or videos)`
state_admin_allotment_report

district_admin_allotment_report
block_admin_allotment_report

### CODE REVIEW

File: modules/rte_mis_allocation/rte_mis_allocation.links.menu.yml [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+#   title: 'RTE MIS Allocation Settings'
+#   route_name: rte_mis_allocation.settings
+#   parent: system.admin_config
+
+# School overview page.
+# rte_mis_allocation.report:
+#   title: 'Student Report'
+#   route_name: rte_mis_allocation.controller.student_admission_report
+#   parent: system.admin_config
+#   description: 'Student Admission Report'
```
Comment (jaswinsingh) [inline modules/rte_mis_allocation/rte_mis_allocation.links.menu.yml:12]: if this is not needed, we can remove this

File: modules/rte_mis_allocation/src/Controller/StudentAdmissionReportController.php [hunk 2]
```diff
@@ -0,0 +1,721 @@
...
+        'Total Not Admitted', 'Total Dropped Out',
+      ];
+    }
+    elseif (in_array('block_admin', $currentUserRole)) {
+      $header = ['No.', 'Total Schools', 'Total RTE seats',
+        'Total Applications', 'Total Applied', 'Total Duplicate', 'Total Incomplete',
+        'Total Rejected', 'Total Approved', 'Total Allotted', 'Total Unallotted', 'Total Admitted',
+        'Total Not Admitted', 'Total Dropped Out',
+      ];
+    }
```
Comment (jaswinsingh) [inline modules/rte_mis_allocation/src/Controller/StudentAdmissionReportController.php:206]: Please add the string in translatable method.

File: modules/rte_mis_allocation/src/Controller/StudentAdmissionReportController.php [hunk 3]
```diff
@@ -0,0 +1,721 @@
...
+          $total_applications, $total_applied, $total_duplicate, $total_incomplete,
+          $total_rejected, $total_approved, $total_allotted, $total_unallotted, $total_admitted, $total_not_admitted, $total_dropout,
+        ];
+        $serialNumber++;
+      }
+
+      return $data;
+    }
+    // Return a markup about missing location.
+    return [['#markup' => $this->t('No districts to display.')]];
```
Comment (jaswinsingh) [inline modules/rte_mis_allocation/src/Controller/StudentAdmissionReportController.php:286]: Can we return single array and not an array within array.

File: modules/rte_mis_allocation/src/Controller/StudentAdmissionReportController.php [hunk 4]
```diff
@@ -0,0 +1,721 @@
...
+            $total_applications, $total_applied, $total_duplicate, $total_incomplete,
+            $total_rejected, $total_approved, $total_allotted, $total_unallotted, $total_admitted, $total_not_admitted, $total_dropout,
+          ];
+          $serialNumber++;
+        }
+        return $data;
+      }
+
+    }
+    return [['#markup' => $this->t('Please check your location.')]];
```
Comment (jaswinsingh) [inline modules/rte_mis_allocation/src/Controller/StudentAdmissionReportController.php:343]: Same here.

File: modules/rte_mis_allocation/src/Controller/StudentAdmissionReportController.php [hunk 5]
```diff
@@ -0,0 +1,721 @@
...
+        $data[] = [$serialNumber, $school_miniNode->get('field_school_name')->getString(), $total_rte_seats,
+          $total_applications, $total_applied, $total_duplicate, $total_incomplete,
+          $total_rejected, $total_approved, $total_allotted, $total_unallotted, $total_admitted, $total_not_admitted, $total_dropout,
+        ];
+      }
+
+      return $data;
+    }
+    // Return a markup about missing location.
+    return [['#markup' => $this->t('Please check your location.')]];
```
Comment (jaswinsingh) [inline modules/rte_mis_allocation/src/Controller/StudentAdmissionReportController.php:401]: Same here

File: modules/rte_mis_allocation/src/Controller/StudentAdmissionReportController.php [hunk 6]
```diff
@@ -0,0 +1,721 @@
...
+        foreach ($locationTree as $value) {
+          if ($value->tid == $id) {
+            return AccessResult::allowed()->setCacheMaxAge(0);
+          }
+        }
+      }
+
+      return AccessResult::allowedIf($id == NULL)->setCacheMaxAge(0);
+    }
+    return AccessResult::allowed()->setCacheMaxAge(0);
```
Comment (jaswinsingh) [inline modules/rte_mis_allocation/src/Controller/StudentAdmissionReportController.php:105]: When are not returning access denied?

Comment (RohanSingh-inno) [reply modules/rte_mis_allocation/src/Controller/StudentAdmissionReportController.php:105]: we are returning access denied if a district / block, tries to access its parents id or same level location id.For example district A tries to access the data of district B. Or if a block tries to access a districts data.

File: modules/rte_mis_allocation/src/Controller/StudentAdmissionReportController.php (hunk 5 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_allocation/src/Controller/StudentAdmissionReportController.php]: ok

File: modules/rte_mis_allocation/src/Controller/StudentAdmissionReportController.php (hunk 4 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_allocation/src/Controller/StudentAdmissionReportController.php]: ok

File: modules/rte_mis_allocation/src/Controller/StudentAdmissionReportController.php (hunk 3 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_allocation/src/Controller/StudentAdmissionReportController.php]: ok

File: modules/rte_mis_allocation/src/Controller/StudentAdmissionReportController.php (hunk 2 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_allocation/src/Controller/StudentAdmissionReportController.php]: ok

File: modules/rte_mis_allocation/rte_mis_allocation.links.menu.yml (hunk 1 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_allocation/rte_mis_allocation.links.menu.yml]: ok

**Review by jaswinsingh — COMMENTED** [review]
@kunal-sin Currently, we have developed the feature where it is mandatory to have district/block admin account, to view a particular report. if admin is not created then it will not show the data under that district/block.
CC - @RohanSingh-inno

**Comment by kunal-sin** [comment]
> @kunal-sin Currently, we have developed the feature where it is mandatory to have district/block admin account, to view a particular report. if admin is not created then it will not show the data under that district/block. CC - @RohanSingh-inno

States should be able to view the district's information, right? Even if the district admin is not present?

**Review by jaswinsingh — APPROVED** [review]

**Pull request merged**